### PR TITLE
fix(gms): rebind non-parameter tensors to private memory after publish

### PR DESCRIPTION
**Pull request by galletas1712**
## Summary

Fixes fatal `NVRM Xid 31` (`FAULT_RO_VIOLATION ACCESS_TYPE_VIRT_WRITE`) GPU faults when a GMS **publisher** engine writes to non-parameter tensors after publish — first hit in practice on all ranks of a Nemotron NVFP4-MoE (fp8 KV, DP8) deployment at the first wake after snapshot restore. The same fault is latent for plain sleep/wake on any `--kv-cache-dtype fp8` model.

### The problem

The publisher builds the entire model inside the GMS `weights` memory pool, so the committed allocations contain more than parameters: torch buffers (`expert_map`, Mamba `conv_weights`, ...) and tensor attributes (fp8 KV scales `_k_scale`/`_v_scale`/`_q_scale`/`_prob_scale`, quantization ranges) are swept into the same caching-allocator segments. That sweep is intentional — a fresh importer never runs `load_weights`, so everything the forward pass references must be materializable from GMS.

The bug is that publish semantics are uniform: after `finalize_gms_write` (and after any RO reconnect — wake, restore), the whole tag is remapped `CU_MEM_ACCESS_FLAGS_PROT_READ`. But some captured tensors are written **after** load:

- `post_kv_cache_wake_up` → `init_fp8_kv_scales` does `fill_()` on the fp8 scale attrs on every wake;
- fp8 range attrs are updated on the forward path.

A write through a `PROT_READ` mapping is a fatal MMU fault. Cold engines only avoid it by timing: every write lands inside the RW load window, and the late writers first run after a sleep/wake — or after a snapshot restore, which is how this was found. Fault forensics: Xid 31 on all ranks at first post-restore wake, fault addresses at the tail 4 KiB page of 2 MiB small-pool segments (fp8 scale blocks at the segment tail).

### The fix

`materialize_module_from_gms` already gives **importers** the right binding: parameters bind to the shared read-only mapping; buffers/tensor attrs are `detach().clone()`d into ordinary CUDA memory. This PR gives the publisher the identical rule:

- `rebind_nonparameter_tensors` (new, `client/torch/module.py`): re-binds every GMS-resident non-parameter tensor (buffers, tensor attrs, list/tuple elements) on the publisher's model to a private `detach().clone()`; parameters stay on the shared read-only mapping. Returns bytes rebound.
- `finalize_gms_write` calls it after `register_module_tensors` + commit + RO remap (so GMS metadata still records the in-pool copies and importer materialization is unchanged), and logs the rebound MiB so per-model duplication stays visible.

Consequences:

- late writers hit ordinary CUDA memory — preserved byte-exact and VA-stable by cuda-checkpoint/CRIU in the snapshot flow;
- the weights tag is honestly `PROT_READ` for every consumer, including snapshot-restored publishers;
- the server's committed copy can never be mutated.

### Notes / constraints

- Runs before CUDA graph capture (clones live at new addresses; captured graphs bake raw pointers). `finalize_gms_write` executes at model-load time in all three integrations (vLLM/TRT-LLM/SGLang), which precedes capture.
- Classification is deliberately blanket — every non-parameter is treated as mutable, because proving immutability per-buffer is impractical. The cost is the clone bytes, logged at finalize.
- Known residual caveat (same semantics importers already have): two attribute names aliasing one tensor become two independent clones after rebind.
- The medium-term direction (a `misc` pool for mutable non-parameter state, converged with KV sleep/wake semantics) is deliberately **not** in this PR; it will be proposed separately as an issue/DEP rather than a checked-in design doc.

## Validation

- New end-to-end regression test `test_finalize_gms_write_rebinds_nonparameter_tensors` in `lib/gpu_memory_service/tests/test_torch_integration.py`: drives the real `finalize_gms_write` flow (register → commit → RO reconnect → remap → rebind) against a live GMS server, asserts the parameter keeps its shared GMS binding while buffer/tensor-attr move to private memory with values preserved, then **writes to the rebound tensors after the RO remap** — the exact operation that faulted.
  - Verified it fails (`assert not _in_gms(gms_model.scale)`) when the `rebind_nonparameter_tensors` call is stubbed out, and passes with the fix.
  - `python -m pytest lib/gpu_memory_service/tests/test_torch_integration.py` → 5 passed on a 2×RTX 5880 Ada box.
- `pre-commit run --all-files` → all hooks pass (black/isort/flake8/ruff/pytest-marker-report).
- Stack-level E2E (branch `schwinns/vllm-snapshot-nightly-overlay-gms`): post-restore inference returns 200 with `finish_reason=stop`, coherent output, no Xid 31, restored weights left read-only. This commit is patch-equivalent to experiment commit `8ef1ab30fb` from that validated image.

AI assistance was used for this PR (review-feedback cleanup and test restructuring); all changes reviewed by the author.

### CODE REVIEW

File: lib/gpu_memory_service/tests/test_module_rebind.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        tensor.data_ptr(): SimpleNamespace(
+            aligned_size=tensor.numel() * tensor.element_size()
+        )
+        for tensor in tensors
+    }
+    return SimpleNamespace(mappings=mappings)
+
+
+@requires_cuda
+def test_rebind_moves_gms_resident_nonparameters_only():
```
[the remark this answers is not in the document]
Comment (galletas1712) [reply lib/gpu_memory_service/tests/test_module_rebind.py]: Fixed in d782a99cc — the standalone test file is gone entirely. The rebind coverage now lives in `test_torch_integration.py` as an end-to-end test (`test_finalize_gms_write_rebinds_nonparameter_tensors`) against a live GMS server, so it inherits that module's `pytestmark` (`pre_merge`, `integration`, `none`, `gpu_1`) and is selected by CI.

File: lib/gpu_memory_service/tests/test_module_rebind.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    model._k_scale.fill_(4.0)
+    assert model._k_scale.item() == 4.0
+
+    # Tensors outside the GMS mappings are already private and untouched.
+    assert model.scales[1].data_ptr() == untracked_list_ptr
+    assert model.private_scale.data_ptr() == private_ptr
+
+
+@requires_cuda
+def test_rebind_is_idempotent_after_clones_leave_gms():
```
[the remark this answers is not in the document]
Comment (galletas1712) [reply lib/gpu_memory_service/tests/test_module_rebind.py]: Fixed in d782a99cc — the second (idempotency) test was folded away along with the file; the surviving end-to-end test in `test_torch_integration.py` carries the module-level markers.

File: lib/gpu_memory_service/client/torch/module.py [hunk 3]
```diff
@@ -221,3 +221,64 @@ def materialize_module_from_gms(
...
+                continue
+            mod[int(attr)] = clone
+        else:
+            if isinstance(getattr(type(mod), attr, None), property):
+                # Read-only derived attribute; the underlying tensor is
+                # iterated (and rebound) separately.
+                logger.debug("[GMS] Skipping property attribute %r", name)
+                continue
+            setattr(mod, attr, clone)
+        rebound_bytes += clone.numel() * clone.element_size()
```
[the remark this answers is not in the document]
Comment (galletas1712) [reply lib/gpu_memory_service/client/torch/module.py]: Fixed in d782a99cc — the loop was restructured so the `detach().clone()` only happens on branches that actually bind it; the skip paths (non-rebindable container, property attribute) no longer allocate.

File: lib/gpu_memory_service/tests/test_module_rebind.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+from types import SimpleNamespace
+
+import pytest
+
+torch = pytest.importorskip("torch")
+
+from gpu_memory_service.client.torch.module import (  # noqa: E402
+    rebind_nonparameter_tensors,
+)
```
[the remark this answers is not in the document]
Comment (galletas1712) [reply lib/gpu_memory_service/tests/test_module_rebind.py]: Resolved in d782a99cc by moving the coverage into `test_torch_integration.py`, which already uses the `_deps` (`HAS_CUDA`/`HAS_GMS`/`HAS_TORCH`) module-level skip pattern.

File: lib/gpu_memory_service/client/torch/module.py [hunk 5]
```diff
@@ -221,3 +221,64 @@ def materialize_module_from_gms(
...
+            tensor_type == "buffer"
+            and hasattr(mod, "_buffers")
+            and attr in mod._buffers
+        ):
+            mod._buffers[attr] = clone
+        elif attr.isdigit() and not isinstance(mod, torch.nn.Module):
+            # Element of a tensor list/tuple attribute.
+            if not isinstance(mod, list):
+                logger.debug("[GMS] Cannot rebind immutable container %r", name)
+                continue
```
[the remark this answers is not in the document]
Comment (galletas1712) [reply lib/gpu_memory_service/client/torch/module.py:285]: Good catch — fixed in d782a99cc. `_iter_module_tensors` does yield tuple elements, so a GMS-resident tensor inside a tuple attribute would have stayed on the PROT_READ mapping (tuple immutability doesn't prevent in-place writes to the tensor itself). Tuples are now rebuilt on their owning attribute, with a property guard on the owner so a derived container can't crash `setattr`. Note the resolution happens fresh per element, so multi-element tuples rebind correctly too.

File: lib/gpu_memory_service/tests/test_module_rebind.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+requires_cuda = pytest.mark.skipif(
+    not torch.cuda.is_available(), reason="requires CUDA"
+)
```
[the remark this answers is not in the document]
Comment (galletas1712) [reply lib/gpu_memory_service/tests/test_module_rebind.py]: Fixed in d782a99cc — the standalone file was removed and the rebind coverage moved into `test_torch_integration.py`, inheriting its `pytestmark` (`pre_merge`, `integration`, `none`, `gpu_1`). (`gpu_0` wouldn't work here: `_iter_module_tensors` only yields CUDA tensors, so the test inherently needs a GPU.) The `pytest-marker-report` pre-commit hook passes.

**Comment by galletas1712** [comment]
Pushed d782a99cc addressing all review feedback:

- **Tuple containers** (CodeRabbit): now rebound by rebuilding the tuple on its owning attribute — previously a GMS-resident tensor in a tuple stayed on the read-only mapping.
- **Clone-before-skip-checks** (Devin): clone is now only allocated on branches that bind it.
- **Test markers / `_deps` pattern** (Devin, CodeRabbit): the standalone fake-manager test file is gone. Coverage moved into `test_torch_integration.py` as a true end-to-end test that drives `finalize_gms_write` against a live GMS server and then writes to the rebound buffer/tensor-attr after the RO remap — the exact operation that faulted with Xid 31. Verified locally that the test fails when the rebind call is stubbed out, and passes with it (5/5 in the file on GPU).
- **Design doc**: dropped `docs/mutable-state-and-memory-classes.md`; the problem/fix rationale now lives in the PR description, and the medium-term misc-pool plan will be filed separately as an issue rather than a checked-in doc that would go stale.
- **`rebound_bytes` in `GMSCommittedMemoryStats`** (CodeRabbit nitpick): declining for now — no caller consumes it and the finalize log line keeps the duplication observable per model; happy to add the field when a metrics consumer exists.
- `pre-commit run --all-files` passes (this was the red CI job — black reformats).

**Review by hhzhang16 — APPROVED** [review]
